### ISSUE TITLE
Could you provide an example of loop break with dependent value?

### ISSUE THREAD

**Issue by NekoiNemo** [danielhenrymantilla/polonius-the-crab.rs#7]
Unlike examples for the `polonius!()`, and the `polonius_loop!()` where the dependent value is returned from function immediately, the suggestion on the [polonius_break!()](https://docs.rs/polonius-the-crab/latest/polonius_the_crab/macro.polonius_break.html) page seem to be rather vague. A naive attempt to implement it results in getting either a mismatched lifetimes error, or the same double borrow error.

Could you please provide a full example for this use case, similar to the examples for the other ones?

**Comment by danielhenrymantilla** [comment]
Hey, yeah I should improve the documentation a bit. Now, the polonius pattern that this crate _currently_ tackles is the one where the _dependent_ value is _returned_. So, indeed, **breaking with a _dependent_ value is currently not supported**.

I can try to work on adding support for it, but it would greatly **help me to see a situation where lack-of-polonius causes problems, with a need-to-break-with-dependent-value situation**. If I had to guess:

  - `break` would correspond to what is our current `return`;
  - `continue` would be the other default branch.

If I'm right about it, and while investigating (the following is not a long-term / proper solution, but if it works it would give me valuable info), could you try to see if wrapping the loop in an _immediately called closure_ would work?

```diff
- polonius_loop! {
+ (|| polonius_loop! {
      |var| -> …'polonius… {
          …
      }
+ })()
- }
```

**Comment by NekoiNemo** [comment]
Maybe i did something wrong, but it didn't work for me - same lifetime error:
```
error[E0495]: cannot infer an appropriate lifetime for lifetime parameter in function call due to conflicting requirements
  --> src/xml2.rs:46:17
   |
46 | /                 polonius_loop!(|buf| -> Option<XmlTag<'polonius>> {
47 | |                     buf.clear();
48 | |
49 | |                     match self.reader.read_event(buf).unwrap() {
...  |
64 | |                     }
65 | |                 })
   | |__________________^
   |
note: first, the lifetime cannot outlive the lifetime `'_` as defined here...
  --> src/xml2.rs:45:23
   |
45 |             let tag = (|| {
   |                       ^^^
note: ...so that closure can access `buf`
  --> src/xml2.rs:46:17
   |
46 | /                 polonius_loop!(|buf| -> Option<XmlTag<'polonius>> {
47 | |                     buf.clear();
48 | |
49 | |                     match self.reader.read_event(buf).unwrap() {
...  |
64 | |                     }
65 | |                 })
   | |__________________^
note: but, the lifetime must be valid for the anonymous lifetime defined here...
  --> src/xml2.rs:41:19
   |
41 |     pub fn root_1(&mut self) -> XmlTag {
   |                   ^^^^^^^^^
note: ...so that the types are compatible
  --> src/xml2.rs:68:13
   |
68 |             tag
   |             ^^^
   = note: expected `xml2::XmlTag<'_>`
              found `xml2::XmlTag<'_>`
   = note: this error originates in the macro `$crate::polonius` (in Nightly builds, run with -Z macro-backtrace for more info)
```

The code is:
```
pub fn root_1(&mut self) -> XmlTag {
        if self.state == XmlState::Init {
            let mut buf = &mut self.buf;

            let tag = (|| {
                polonius_loop!(|buf| -> Option<XmlTag<'polonius>> {
                    buf.clear();

                    match self.reader.read_event(buf).unwrap() {
                        Event::Start(bs) => polonius_return!(Some(XmlTag {
                            data: bs,
                            is_empty: false
                        })),
                        Event::Empty(bs) => polonius_return!(Some(XmlTag {
                            data: bs,
                            is_empty: true
                        })),
                        Event::Eof => polonius_break!(None),
                        Event::End(ev) => panic!(
                            "Encountered </{}> before reaching the root of XML document",
                            self.reader.decode(ev.name()).unwrap()
                        ),
                        _ => (),
                    }
                })
            })().unwrap();

            tag
        } else {
            panic!("Reader isn't fresh - can't read to root")
        }
    }
```

Previously i solved the problem by putting what is currently in the anonymous function into a separate named function and using `poloniuous_return!()` from it, and calling it in `let tag = ...`

**Comment by danielhenrymantilla** [comment]
Hmm yeah my bad the `polonius_break!` would have to be replaced with `polonius_return!` (the whole point of the closure wrapping dance), since _currently_ only `polonius_return!` allows to return a dependent type.

The following does work, for instance:

```rust
pub fn root_1(&mut self) -> XmlTag {
    if self.state == XmlState::Init {
        let reader = &mut self.reader;

        let tag = (move |mut buf| {
            let _: &mut Vec<u8> = buf;
            polonius_loop!(|buf| -> Option<XmlTag<'polonius>> {
                buf.clear();

                match reader.read_event(buf).unwrap() {
                    Event::Start(bs) => polonius_return!(Some(XmlTag {
                        data: bs,
                        is_empty: false
                    })),
                    Event::Empty(bs) => polonius_return!(Some(XmlTag {
                        data: bs,
                        is_empty: true
                    })),
                    Event::Eof => polonius_return!(None),
                    Event::End(ev) => panic!(
                        "Encountered </{}> before reaching the root of XML document",
                        reader.decode(ev.name()).unwrap()
                    ),
                    _ => (),
                }
            })
        })(&mut self.buf)
        .unwrap();

        tag
    } else {
        panic!("Reader isn't fresh - can't read to root")
    }
}
```

  - [Demo](https://www.rustexplorer.com/b/y0v1cq).

**Comment by danielhenrymantilla** [comment]
Now, I have to admit that this does showcase **something missing in the API of polonius**. For what is worth, here is what the direct usage of the advanced function (no macros) would look like:

```rust
let mut buf = &mut self.buf;
let tag = loop {
    /// `type Return = HKT!(<'polonius> => Option<XmlTag<'polonius>>);`
    type Return = dyn for<'p> WithLifetime<'p, T = Option<XmlTag<'p>>>;
    use Ok as Break;
    match polonius::<Return, _, _, _>(buf, |buf| {
        buf.clear();
        match reader.read_event(buf).unwrap() {
            Event::Start(bs) => Break(Some(XmlTag {
                data: bs,
                is_empty: false
            })),
            Event::Empty(bs) => Break(Some(XmlTag {
                data: bs,
                is_empty: true
            })),
            Event::Eof => Break(None),
            Event::End(ev) => panic!(
                "Encountered </{}> before reaching the root of XML document",
                reader.decode(ev.name()).unwrap()
            ),
            _ => Err(()),
        }
    })
    {
        Break(it) => break it,
        Err((buf_, ())) => buf = buf_,
    }
};
tag.unwrap()
```

  - [Demo](https://www.rustexplorer.com/b/c8ba6c)

For what is worth, that outer `match` unsugaring (the one with the `break it`) is _exactly_ like `polonius!`'s, but for it using `break` instead of `return`. This thus gives an idea regarding how to implement your desired API: what would you think of:

```rust
let mut buf = &mut self.buf;
let tag = loop {
    polonius!(|buf| -> Option<XmlTag<'polonius>> {
        buf.clear();
        match reader.read_event(buf).unwrap() {
            Event::Start(bs) => polonius_return!(Some(XmlTag {
                data: bs,
                is_empty: false
            })),
            Event::Empty(bs) => polonius_return!(Some(XmlTag {
                data: bs,
                is_empty: true
            })),
            Event::Eof => polonius_return!(None),
            Event::End(ev) => panic!(
                "Encountered </{}> before reaching the root of XML document",
                reader.decode(ev.name()).unwrap()
            ),
            _ => (),
        }
    }, return = break)
};
tag.unwrap()
```

And maybe I could also tweak `polonius_loop!` to offer something similar 🤔

**Comment by NekoiNemo** [comment]
Thank you.
As for the API - i think that might be a bit confusing, with redefining return as break at the end, and using the already existing `polonius_return` to now have a different function than being a return.

**Comment by danielhenrymantilla** [comment]
Yeah, you're right. After thinking about it, I've realized there is a (complex for me to implement) way to integrate this into `polonius_loop!`, while keeping all the other macros working. The end result with this API, tested on my machine would be:

  - ```rust
    pub fn root_1(&mut self) -> XmlTag {
        if self.state == XmlState::Init {
            let mut buf = &mut self.buf;
            let tag = polonius_loop!(|buf| -> _, break: Option<XmlTag<'polonius>> {
                buf.clear();
                polonius_break_dependent!(match self.reader.read_event(buf).unwrap() {
                    Event::Start(bs) => Some(XmlTag {
                        data: bs,
                        is_empty: false
                    }),
                    Event::Empty(bs) => Some(XmlTag {
                        data: bs,
                        is_empty: true
                    }),
                    Event::Eof => None,
                    Event::End(ev) => panic!(
                        "Encountered </{}> before reaching the root of XML document",
                        self.reader.decode(ev.name()).unwrap()
                    ),
                    _ => polonius_continue!(),
                })
            })
            .unwrap();

            tag
        } else {
            panic!("Reader isn't fresh - can't read to root")
        }
    }
    ```

  - I may even be able to conflate `polonius_break!` and `polonius_break_dependent!` with the only difference being whether that `, break: …` extra type annotation in the "closure header" has been added or not 🤔 

I'll clean it up and release it, which ought to be a fix tackling this very thread 🙂

**Comment by danielhenrymantilla** [comment]
Fixed by #8 (see https://docs.rs/polonius-the-crab/0.3.0-rc2/polonius_the_crab/macro.polonius_loop.html#breaks)
